### ISSUE TITLE
Typings: allow including only used components in app bundle

### ISSUE THREAD

**Issue by PavelPZ** [Semantic-Org/Semantic-UI-React#1080]
When I use
```import Button from 'semantic-ui-react/elements/Button/Button'```
instead of
```import { Button } from 'semantic-ui-react';```
I can save almost half of the project's bundle.js file.

See example of Button and Accordion definition in https://github.com/PavelPZ/design/blob/master/index.d.ts.
which allows this optimization.

------------------------------
**APP Code:**
```
import React from 'react';
import ReactDOM from 'react-dom';

//OPTIMIZED VARIANT, bundle contains only Button code
import Button from 'semantic-ui-react/elements/Button/Button';
import ButtonContent from 'semantic-ui-react/elements/Button/ButtonContent';

//STANDARD VARIANT, bundle contains all Semantic UI code
//import { Button, ButtonContent } from 'semantic-ui-react'; //STANDARD VARIANT

export const init = () => {
  ReactDOM.render(<Button primary><ButtonContent>Button</ButtonContent></Button>, document.getElementById('content'));
};
```

**Minified bundle size (containing APP Code, react.js, lodash.js and classnames.js):**

OPTIMIZED VARIANT: 326kB
STANDARD VARIANT: 592kB

Used JSPM package manager with Rollup Optimization.

**Comment by cschmidli** [comment]
another option would be to have an index.d.ts in each component folder, which is similar to the typescript compiler approach. E.g. react toolbox follows that approach: https://github.com/react-toolbox/react-toolbox/tree/dev/components/card.

I guess the nice part is that you don't end up with a 5000 lines typings file :)

**Comment by PavelPZ** [comment]
@sagax85, your approach is more common when Typescript is supported by library authors.

See https://github.com/Semantic-Org/Semantic-UI-React/issues/361 or https://github.com/Semantic-Org/Semantic-UI-React/pull/674.

**Comment by levithomason** [comment]
I'm all for smaller bundles.  I am not, however, a typescript user.  How does this approach square with https://github.com/Semantic-Org/Semantic-UI-React/pull/674#issuecomment-265295718?

Also, if you're looking for bundle optimization you should also have a look at https://github.com/Semantic-Org/Semantic-UI-React/pull/731#issuecomment-258729112.

**Comment by levithomason** [comment]
@asvetliakov what is your input on the refactor proposed in this issue?

**Comment by asvetliakov** [comment]
@levithomason 
You can have separate typings for each component, i.e. similar to this structure:
```
 --elements
    index.js
    index.d.ts
     --button
        button.js
        button.d.ts
     --label
        label.js
        label.d.ts
---

button.d.ts
export interface ButtonProps { }
export default class Button extends React.Component<...> { }

elements/index.js and index.d.ts will be same
export { default as Button } from "./button/button";
```

This is easy done when you're writing in typescript, since definitions will be autogenerated, but will require some manual management for JS based project.

Other libs sometime include separate ambient module definitions for substructure modules , i.e. ```declare module "lib/elements/button" { ... } ``` but it's allowed only for ```@types```.

So your options are:
1) Just tell your users to use ES6 tree-shaking (rollup and webpack2 have this ability, but i'm not sure about tree shaking from node_modules includes though)
2) Split your typing file as in my example above
3) Remove typings file and send PR to DT repo. This way you can maintain single file with multiple submodule definitions

**Comment by levithomason** [comment]
Thanks for the in-depth feedback.  I'll look to the TS users here for preferences.

**Comment by cschmidli** [comment]
I vote for option 2 (separate files), because:
* nicer to maintain as the files are kept clean and are stored just next to where the component/JS code is
* easier even for non TS users to adjust the typings file in case of changes to an existing component
* works really well with the auto import plugins for VSCode/WebStorm etc.
* authors don't seem to be against supporting typescript

**Comment by levithomason** [comment]
Option 2 it is, PRs welcome!

**Comment by cschmidli** [comment]
ok i'll work on a PR unless someone else already started?

**Comment by layershifter** [comment]
Hm, as I see this was fixed in #1098. Closing for housekeeping.
